### ISSUE TITLE
C# wrapper for SAT solver

### ISSUE THREAD

**Issue by teocomi** [google/or-tools#552]
Question:
Is the C# wrapper for the SAT solver planned to be released any time soon?

Context:
I have a python script working very nicely with the SAT solver but need to integrate it into a C# application.
After reading #417 I followed the example in examples/test/testsat.cs to convert my python code, but when I add several variables to the linear constraint I get the error `Unable to parse Google.OrTools.Sat.CpSolverResponse protocol message.`, it seems to work less variables.
Also I don't know how search for all the solutions in C#.

**Comment by Mizux** [comment]
Hi, I won't say soon but definitively in the pipe !

Currently we only have a "minimal" swig csharp wrapper, but like in python with its custom `cp_model.py`, having a custom csharp interface will be worth it IMHO.

Also I need to find a way to maintain a better API consistency between languages (i.e could imply breaking API to have a "better" cross language consistency/coherency).

**Comment by Mizux** [comment]
Hi,

Good news, @lperron has worked on C# SAT Solver and started to port most examples on it...
Please fetch the last master revision and give it a try ;)

**Comment by dhanarfmlx** [comment]
Hi,

I would like to tell you that I am still encountering the same error message:  Unable to parse Google.OrTools.Sat.CpSolverResponse protocol message. I also don't encounter the error if I use less variables. 

I am using Google.OrTools.6.7.4973. Is this a known issue? Is there any known workaround for this?

Thanks.

**Comment by dhanarfmlx** [comment]
Hi,

Attached with this comment is the code sample that exhibits the problem. It is a VS2012 solution. Pardon me if the code is messy because I am still learning.

Thanks you.

[JspPm_ParseError.zip](https://github.com/google/or-tools/files/1922035/JspPm_ParseError.zip)

**Comment by lperron** [comment]
I ran from source,

Horizon: 75

Statistics

  - solve status : Optimal

  - makespan     : 25

  - conflicts    : 45

  - branches     : 547

  - wall time    : 0.012902394 ms


Protocol 0 Task 0 Machine 100 Start 0 End 3 Duration 3

Protocol 0 Task 1 Machine 201 Start 3 End 4 Duration 1

Protocol 0 Task 2 Machine 301 Start 4 End 6 Duration 2

Protocol 0 Task 3 Machine 201 Start 6 End 10 Duration 4

Protocol 0 Task 4 Machine 100 Start 10 End 15 Duration 5

Protocol 1 Task 0 Machine 101 Start 1 End 4 Duration 3

Protocol 1 Task 1 Machine 201 Start 5 End 6 Duration 1

Protocol 1 Task 2 Machine 301 Start 7 End 9 Duration 2

Protocol 1 Task 3 Machine 200 Start 9 End 13 Duration 4

Protocol 1 Task 4 Machine 101 Start 13 End 18 Duration 5

Protocol 2 Task 0 Machine 101 Start 10 End 13 Duration 3

Protocol 2 Task 1 Machine 200 Start 13 End 14 Duration 1

Protocol 2 Task 2 Machine 301 Start 14 End 16 Duration 2

Protocol 2 Task 3 Machine 200 Start 16 End 20 Duration 4

Protocol 2 Task 4 Machine 101 Start 20 End 25 Duration 5

Protocol 3 Task 0 Machine 101 Start 4 End 7 Duration 3

Protocol 3 Task 1 Machine 200 Start 7 End 8 Duration 1

Protocol 3 Task 2 Machine 301 Start 9 End 11 Duration 2

Protocol 3 Task 3 Machine 201 Start 11 End 15 Duration 4

Protocol 3 Task 4 Machine 100 Start 15 End 20 Duration 5

Protocol 4 Task 0 Machine 101 Start 7 End 10 Duration 3

Protocol 4 Task 1 Machine 201 Start 10 End 11 Duration 1

Protocol 4 Task 2 Machine 301 Start 11 End 13 Duration 2

Protocol 4 Task 3 Machine 201 Start 15 End 19 Duration 4

Protocol 4 Task 4 Machine 100 Start 20 End 25 Duration 5

Therefore, I do not know the source of the bug.
I suspect a protobuf version mismatch. We will investigate.



Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by dhanarfmlx** [comment]
@lperron It is strange that you can run the code in **JspPm_ParseError.zip** without any error. 

I have double-checked that the code is still causing problem on my laptop. By the way, I am on Windows 10 64-bit latest update as of today with Intel Core i7 7500U and 12 GB of RAM if that helps.

Try to increase the number of data, by doubling the loop count and see if you can reproduce the issue.

Replace

`for (var i = 0; i < 5; i++) // this is having problem`

With

`for (var i = 0; i < 10; i++) // this is having problem`

If you suspect that it is protobuf mismatch, then probably there is a problem with or-tools package in Nuget.

**Comment by Haukinger** [comment]
I'm experimenting with the job-shop-sample ([link](https://github.com/google/or-tools/blob/master/examples/csharp/jobshop_ft06_sat.cs)), and until now, I just increased the number of jobs and machines looking for the limitations of the solver. With 15 machines and 9 jobs I get the error.

I've tried both protobuf 3.5.1 as in the nuget package and the current 3.6.0 version. Windows 7 x64 on i7-3520M here.
